### PR TITLE
feat: Update useTheme hook

### PR DESCRIPTION
**Pull request by RayRedGoose**
## Summary

Fixes: #923 

- We've added `getTheme` so that you can have access to theme inside of `styled` function or class components.
- We've updated `useTheme` to add a warning if being used outside of a context. 
- We've removed `getCanvasTheme` and `useCanvasTheme`.

## Release Category
Components

### BREAKING CHANGES
- We've removed `getCanvasTheme` and `useCanvasTheme`.
- Use `useTheme` when you have a `CanvasProvider` or you're inside of a functional component
- Use `getTheme` when inside a styled component and you need access to theme.

---

## Checklist

- [ ] MDX documentation adheres to Canvas Kit's [standard MDX template](https://github.com/Workday/canvas-kit/discussions/1131)
- [ ] Label `ready for review` has been added to PR

## For the Reviewer

- [ ] PR title is short and descriptive
- [ ] PR summary describes the change (Fixes/Resolves linked correctly)
- [ ] PR Release Notes describes additional information useful to call out in a release message or removed if not applicable
- [ ] Breaking Changes provides useful information to upgrade to this code or removed if not applicable

## Areas for Feedback? (optional)

- [ ] Code
- [ ] Documentation
- [ ] Testing
- [ ] Codemods

### CODE REVIEW

**Comment by mannycarrera4** [comment]
Heads up, I believe we have a `getCanvasTheme` that looks like its doing similar work to `getTheme`. 
We should keep the validation in `useTheme` but should we deprecate it?

**Comment by NicholasBoll** [comment]
The contents of `getTheme` and `useTheme` were not what I expected. They contain duplicate implementations. I was thinking `useTheme` made calls to `getTheme` instead of containing duplicated functionality.

File: modules/react/common/lib/theming/README.md [hunk 1]
```diff
@@ -241,6 +241,168 @@ window.workday = {
...
+## Access the theme value
+
+To access the current theme, you can use one of the following functions and hooks depending on where
+do you use it and what do you want to get back. All of these functions allow to pass a partial or
+full theme object to overwrite the current theme. If a theme is not passed, they will try to
+retrieve it from the ThemeContext (only for hooks), then from the window object and as a last
+solution they will fall back to the default theme.
+
+Please consider to use `getTheme` and `useTheme` over `getCanvasTheme` and `useCanvasTheme`.
+
```
[the remark this answers is not in the document]
Comment (alanbsmith) [reply modules/react/common/lib/theming/README.md:253]: I think we want something more like this: `getCanvasTheme` and `useCanvasTheme` use a Proxy, so you trade completeness for performance. If you only need the last property in a member expression, these are a good option.

**Review by alanbsmith — CHANGES_REQUESTED** [review]
The code for this looks good. Thanks, Raisa. 😄 I'd like to tweak the documentation a bit more to help users understand how these four functions are different and when they should choose which option.

File: modules/react/common/lib/theming/index.ts [hunk 2]
```diff
@@ -5,5 +5,4 @@ export {default as styled, StyleRewriteFn} from './styled';
 export * from './theme';
 export * from './useTheme';
 export * from './useIsRTL';
-export * from './getCanvasTheme';
 export * from './getObjectProxy';
```
Comment (mannycarrera4) [inline modules/react/common/lib/theming/index.ts:8]: do we still need objectProxy?  i know this was used for get canvas theme

File: modules/react/common/lib/theming/useTheme.ts [hunk 3]
```diff
@@ -28,26 +80,31 @@ const getFilledTheme = (theme: PartialEmotionCanvasTheme) => ({
...
+    try {
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      const contextTheme = useEmotionTheme() as EmotionCanvasTheme;
+      if (contextTheme?.canvas) {
+        return getFilledTheme(contextTheme);
+      }
+    } catch (e) {
+      if (process && process.env.NODE_ENV === 'development') {
+        console.warn(
+          'useTheme: Context not supported or invalid. Please consider to use `getTheme` function instead for `styled` or class components.'
```
Comment (mannycarrera4) [inline modules/react/common/lib/theming/useTheme.ts:103]:
```suggestion
          'useTheme: Context not supported or invalid. Please consider to using `getTheme` function instead for `styled` or class components.'
```

**Comment by mannycarrera4** [comment]
Code wise this looks good to me! I'll let someone else take a look at the docs! 👍

File: modules/react/common/lib/theming/README.md [hunk 4]
```diff
@@ -241,6 +241,93 @@ window.workday = {
...
+do you use it and what do you want to get back. These functions allow to pass a partial or full
+theme object to overwrite the current theme. If a theme is not passed, they will try to retrieve it
+from the ThemeContext (only for hooks), then from the window object and as a last solution they will
+fall back to the default theme.
+
+### getTheme
+
+`getTheme` is function to get the full theme object. It should be used with `styled` and class
+components or outside components. Function returns a theme object with the Canvas Kit theme under
+the canvas key.
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/README.md:256]:
```suggestion
`getTheme` is a function that returns the full theme object with the Canvas Kit theme under the `canvas` key. It should be used with `styled` components, class components, or outside components.
```

File: modules/react/common/lib/theming/README.md [hunk 5]
```diff
@@ -241,6 +241,93 @@ window.workday = {
...
+    breakpoints: {
+      // ...
+    },
+    direction: ContentDirection.LTR,
+  },
+  // other themes can be placed here
+}
+```
+
+**Example**
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/README.md:275]:
```suggestion
Below is an example of how to use `getTheme` to build responsive media query styles with the breakpoint functions provided in the theme.
```

File: modules/react/common/lib/theming/README.md [hunk 6]
```diff
@@ -241,6 +241,93 @@ window.workday = {
...
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/README.md:320]:
```suggestion
Below is an example showing how to use `useTheme` in a function component to set `Subtext`'s color to the error color provided by the theme.
```

File: modules/react/common/lib/theming/README.md [hunk 7]
```diff
@@ -241,6 +241,93 @@ window.workday = {
...
+    },
+    breakpoints: {
+      // ...
+    },
+    direction: ContentDirection.LTR,
+  },
+  // other themes can be placed here
+}
+```
+
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/README.md:274]: Let's remove this and add it to the top to remove duplication. 😄
```suggestion
```

File: modules/react/common/lib/theming/README.md [hunk 8]
```diff
@@ -241,6 +241,93 @@ window.workday = {
...
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/README.md:319]: Let's remove this and add it to the top to remove duplication. 😄
```suggestion
```

File: modules/react/common/lib/theming/useTheme.ts [hunk 9]
```diff
@@ -12,12 +15,61 @@ const getFilledTheme = (theme: PartialEmotionCanvasTheme) => ({
...
+ *
+ * const styles = {
+ *  [small]: {
+ *    margin: space.m
+ *  },
+ *  [medium]: {
+ *    margin: space.l
+ *  }
+ * }
+ */
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/useTheme.ts:49]: This is really nice JSDoc. Thank you. 😄

File: modules/react/common/lib/theming/useTheme.ts [hunk 10]
```diff
@@ -28,26 +80,31 @@ const getFilledTheme = (theme: PartialEmotionCanvasTheme) => ({
...
  * ThemeProvider or context exists.
  * Tracked on https://github.com/emotion-js/emotion/issues/1193.
+ *
+ * @example
+ * export const ErrorMessage = () => {
+ *  const theme = useTheme();
+ *  return (
+ *    <Subtext size="large" color={theme.canvas.palette.error.main}>
+ *  );
+ * }
```
Comment (alanbsmith) [inline modules/react/common/lib/theming/useTheme.ts:90]: Same here. I love all this detail. It's very clear and readable.

**Review by alanbsmith — APPROVED** [review]
Looks great, Raisa. I made some documentation nits, but overall, this looks good.

File: modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx [hunk 11]
```diff
@@ -14,12 +14,11 @@ any questions.
...
 - [Component Updates](#component-updates)
-  - [Buttons](#buttons)
-	- [Toast](#toast)
+  - [Buttons](#buttons) - [Toast](#toast)
 - [Utility Updates](#utility-updates)
   - [focusRing](#focusring)
-- [Token Updates](#token-updates)
-	- [Depth](#depth)
+  - [Theme hooks](#theme-hook)
+- [Token Updates](#token-updates) - [Depth](#depth)
```
Comment (mannycarrera4) [inline modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx:21]: can you update this, it happens to me all the time

**Review by mannycarrera4 — APPROVED** [review]

File: modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx [hunk 12]
```diff
@@ -175,6 +176,17 @@ We're removing memoization from focus ring. The `memoize` argument passed to `fo
...
 
+### useTheme and getTheme
+
+We've updated `useTheme` by adding error handling if this hook has been used outside a functional
+component. Also, `getTheme` method has been added to access a theme from `styled` or class
+components instead of `useTheme`.
+
+### useCanvasTheme and getCanvasTheme
+
+`useCanvasTheme` and `getCanvasTheme` have been removed as it's can be safely replaced by `useTheme`
```
Comment (mannycarrera4) [inline modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx:187]:
```suggestion
`useCanvasTheme` and `getCanvasTheme` have been removed as they can be safely replaced by `useTheme`
```

File: modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx [hunk 13]
```diff
@@ -175,6 +176,17 @@ We're removing memoization from focus ring. The `memoize` argument passed to `fo
...
+and `getTheme` and there is no external usage of this functions found.
```
Comment (mannycarrera4) [inline modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx:188]:
```suggestion
and `getTheme`.
```

File: modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx (hunk 12 above)
Comment (mannycarrera4) [inline modules/docs/mdx/9.0-UPGRADE-GUIDE.mdx:187]:
```suggestion
`useCanvasTheme` and `getCanvasTheme` have been removed and can be safely replaced by `useTheme`
```

**Review by mannycarrera4 — APPROVED** [review]

**Pull request merged**